internal/core/src/storage/GrowingMmapChunkManager.cpp

```diff
@@ -0,0 +1,243 @@
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/GrowingMmapChunkManager.h"
+#include "storage/LocalChunkManagerSingleton.h"
+#include <fstream>
+#include <sys/mman.h>
+#include <unistd.h>
+#include "stdio.h"
+#include <fcntl.h>
+#include "log/Log.h"
+
+namespace milvus::storage {
+namespace {
+static constexpr int kMmapDefaultProt = PROT_WRITE | PROT_READ;
+static constexpr int kMmapDefaultFlags = MAP_SHARED;
+};  // namespace
+
+MappEntry::MappEntry(const std::string& file_name,
+                     const uint64_t file_size,
+                     EntryType type)
+    : file_name_(file_name), file_size_(file_size), entry_type_(type) {
+    // create tmp file
+    int fd = open(file_name_.c_str(), O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
+    if (fd == -1) {
+        LOG_ERROR("Failed to open mmap tmp file");
+        return;
+    }
+    // append file size to 'file_size'
+    if (lseek(fd, file_size - 1, SEEK_SET) == -1) {
+        LOG_ERROR("Failed to seek mmap tmp file");
+        return;
+    }
+    if (write(fd, "", 1) == -1) {
+        LOG_ERROR("Failed to write mmap tmp file");
+        return;
+    }
+    // memory mapping
+    addr_ = static_cast<char*>(
+        mmap(nullptr, file_size, kMmapDefaultProt, kMmapDefaultFlags, fd, 0));
+    if (addr_ == MAP_FAILED) {
+        LOG_ERROR("Failed to mmap");
+        return;
+    }
+    offset_.store(0);
+}
+
+MappEntry::~MappEntry() {
+    if (addr_ != nullptr) {
+        munmap(addr_, file_size_);
+    }
+    if (std::ifstream(file_name_.c_str()).good()) {
+        std::remove(file_name_.c_str());
+    }
+}
+
+void*
+MappEntry::Allocate(const uint64_t size, ErrorCode& error_code) {
+    if (file_size_ - offset_.load() < size) {
+        error_code = ErrorCode::MemAllocateFailed;
+        return nullptr;
+    } else {
+        error_code = ErrorCode::Success;
+        return (void*)(addr_ + offset_.fetch_add(size));
+    }
+}
+
+MappEntryPtr
+FixedSizeMmapEntryQueue::Pop() {
+    if (queuing_entries_.empty()) {
+        return std::make_unique<MappEntry>(
+            mmap_entry_meta_ptr_->GetMmapFilePath(),
+            mmap_entry_meta_ptr_->GetFixFileSize(),
+            MappEntry::EntryType::NORMAL);
+    } else {
+        std::lock_guard<std::shared_mutex> lock(entry_mtx_);
+        auto entry = queuing_entries_.back();
+        free_disk_size_.fetch_sub(entry->GetCapacity());
+        queuing_entries_.pop();
+        return entry;
+    }
+}
+
+void
+FixedSizeMmapEntryQueue::Push(MappEntryPtr&& entry) {
+    free_disk_size_.fetch_add(entry->GetCapacity());
```

Comment:
If we put this under the scope of the lock, then it has not to be a atomic data type